AdvLoggerPkg/AdvLoggerOsConnectorPrm/Library/AdvLoggerOsConnectorPrmConfigLib/AdvLoggerOsConnectorPrmConfigLib.c

```diff
@@ -0,0 +1,278 @@
+    @return         BOOLEAN     FALSE- LoggerInfo failed security checks
+
+**/
+STATIC
+BOOLEAN
+ValidateInfoBlock (
+  ADV_LOGGER_PRM_DATA_BUFFER  *DataBuf
+  )
+{
+  if ((DataBuf == NULL) || (DataBuf->LoggerInfo == NULL)) {
+    return FALSE;
+  }
+
+  if (DataBuf->LoggerInfo->Signature != ADVANCED_LOGGER_SIGNATURE) {
+    return FALSE;
+  }
+
+  if (DataBuf->LoggerInfo->LogBufferOffset != sizeof (*(DataBuf->LoggerInfo))) {
+    return FALSE;
+  }
+
+  if ((DataBuf->LoggerInfo->LogCurrentOffset > TOTAL_LOG_SIZE_WITH_ALI (DataBuf->LoggerInfo)) ||
+      (DataBuf->LoggerInfo->LogCurrentOffset < DataBuf->LoggerInfo->LogBufferOffset))
+  {
+    return FALSE;
+  }
+
+  // ensure that the passed in sizes match what we see in the advanced logger structure
+  if ((DataBuf->ExpectedLogSize != DataBuf->LoggerInfo->LogBufferSize) ||
+      (DataBuf->ExpectedHeaderSize != DataBuf->LoggerInfo->LogBufferOffset))
+  {
+    return FALSE;
+  }
+
+  return TRUE;
+}
+
+/**
+  Constructor of the PRM configuration library.
+
+  @param[in] ImageHandle        The image handle of the driver.
+  @param[in] SystemTable        The EFI System Table pointer.
+
+  @retval EFI_SUCCESS           The shell command handlers were installed successfully.
+  @retval EFI_UNSUPPORTED       The shell level required was not found.
+**/
+EFI_STATUS
+EFIAPI
+AdvLoggerOsConnectorPrmConfigLibConstructor (
+  IN  EFI_HANDLE        ImageHandle,
+  IN  EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  EFI_STATUS                  Status;
+  PRM_CONFIG_PROTOCOL         *PrmConfigProtocol = NULL;
+  ADVANCED_LOGGER_PROTOCOL    *LoggerProtocol;
+  ADV_LOGGER_PRM_DATA_BUFFER  *DataBuf = NULL;
+  PRM_CONTEXT_BUFFER          *PrmContextBuffer;
+  UINTN                       DataBufferLength;
+
+  //
+  // Length of the data buffer = Buffer Header Size + Size of LoggerInfo pointer
+  //
+  DataBufferLength = sizeof (PRM_DATA_BUFFER_HEADER) + sizeof (ADV_LOGGER_PRM_DATA_BUFFER);
+
+  mStaticDataBuffer = AllocateRuntimeZeroPool (DataBufferLength);
+  if (mStaticDataBuffer == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  //
+  // Locate the Logger Information block.
+  //
+  Status = gBS->LocateProtocol (
+                  &gAdvancedLoggerProtocolGuid,
+                  NULL,
+                  (VOID **)&LoggerProtocol
+                  );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a Failed to find Advanced Logger Protocol\n", __func__));
+    goto Done;
+  }
+
+  //
+  // To ensure we only read the size of the log buffer in the PRM, read the log buffer size
+  // directly from the PCD. We need to share the header size we are using, which is recorded as
+  // the log buffer offset
+  //
+  DataBuf                     = (ADV_LOGGER_PRM_DATA_BUFFER *)mStaticDataBuffer->Data;
+  DataBuf->LoggerInfo         = LOGGER_INFO_FROM_PROTOCOL (LoggerProtocol);
+  DataBuf->ExpectedLogSize    = EFI_PAGES_TO_SIZE (FixedPcdGet32 (PcdAdvancedLoggerPages)) - sizeof (ADVANCED_LOGGER_INFO);

```

Comment:
Is the case `PcdAdvancedLoggerPages == 0` handled gracefully somewhere else? This would lead to unexpected arithmetic results in places like this.

Reply:
No, unfortunately there is a lot of grace that could be added to Advanced Logger :). All the other code does exactly this. What I will do is handle this gracefully here, if the Pcd is < sizeof (AdvancedLoggerInfo), don't publish the config protocol, this is a misconfiguration. I will file an issue to track fixing this gracefully in the code that sets up the log buffer.

Reply:
Tracked in bug #475.